### PR TITLE
 DAOS-6284 Server: Telemetry init and initial metrics

### PR DESCRIPTION
**Pull request by jbrosenz**
Made some improvements to the telemetry library.
    d_tm_list() now operates recursively from the node
    given, rather than provide a list only one level deep.

    Made minor updates to address some Coverity scan errors.

    Jenkinsfile now includes libyaml-devel

    Instrument the following:
      * ioserver
          * startup_duration
          * start_time
          * rank
      * rpc
          * queue stats
            * aborts
            * wait_depth
            * requests
            * replies
            * inflight_depth
      * uri lookups
          * requests
          * cache hit
          * cache miss
 
    Adds daos_metrics utility program to read metrics and display
    a telemetry tree.  Specify the server index, path, and filters to
    customize the output shown.  See daos_metrics --help for
    usage.

### CODE REVIEW

File: src/utils/daos_metrics/daos_metrics.c
Comment (mjmac) [inline src/utils/daos_metrics/daos_metrics.c:194]: I'm not sure this output adds much from a user perspective. The way I would use this is to run the tool using watch, e.g.

`$ watch -n1 daos_metrics -p server -i1`

In this case the iteration count is always 0, but even if a developer isn't using watch, the iteration and matching metrics count just seems to add visual noise to the output.

**Review by mjmac — COMMENTED** [review]
Looks good to me, just need to adjust the output and merge master to get some test updates.

File: src/iosrv/init.c
Comment (mjmac) [inline src/iosrv/init.c:466]: Might be worthwhile to add a metric for the rank as well. If we're looking at an aggregated tree of metrics it would be useful to see the ranks in it. Could also be a way to query for rank via metrics rather than scraping logs or whatever.

File: src/utils/daos_metrics/daos_metrics.c
Comment (jbrosenz) [reply src/utils/daos_metrics/daos_metrics.c:194]: That makes sense to me.  I'll remove those two statements.

File: src/iosrv/init.c
Comment (jbrosenz) [reply src/iosrv/init.c:466]: Done

**Review by mjmac — DISMISSED** [review]
LGTM

File: src/cart/crt_context.c
Comment (johannlombardi) [inline src/cart/crt_context.c:303]: This function is called on the client side too. Could you please elaborate on what happens there?
On the server side, i think we said that this should be reported per target.

File: src/cart/crt_context.c
Comment (johannlombardi) [inline src/cart/crt_context.c:996]: Why not putting all of this in the context instead?

File: src/cart/crt_rpc.c
Comment (johannlombardi) [inline src/cart/crt_rpc.c:758]: same comment here.

**Review by johannlombardi — APPROVED** [review]
Ok to land in order to move the needle, but we will have to revisit the metrics added to the cart code.

**Comment by johannlombardi** [comment]
Adding @frostedcmos for the cart changes.

File: src/cart/crt_context.c
Comment (jbrosenz) [reply src/cart/crt_context.c:303]: In the current implementation, when the client side calls d_tm_increment_counter, the function validates if the T&M is initialized for this instance or not, determines that it is not initialized, and returns -DER_UNINIT.  The caller is not processing these return values, so there is nothing logged.  The return codes are available for a developer to use as they are developing their code.  My expectation is that the examination of the return value would not make it into the production code once the developer verifies that everything is working as it should.

File: src/cart/crt_context.c
Comment (jbrosenz) [reply src/cart/crt_context.c:996]: My intention was that these were global across all contexts.  I see your point that having these local to a context would be helpful.  There's one consideration with that, that I need to work through first.  Is there a fixed number of these contexts for the life of the IO server?  If the number is fixed, then it's straight forward to add these values to the context.  If the number grows over time as contexts come and go, then there's a memory management issue to work through to reclaim the shared memory that was allocated to an old unused context to avoid running out of shared memory as new contexts (with new data sets) are added.  Can you advise about the nature of the context?

**Comment by jbrosenz** [comment]
The one error in testing (as of the current commit) is an instance of DAOS-6507.

File: src/cart/crt_context.c
Comment (mjmac) [reply src/cart/crt_context.c:303]: Hmm. I don't think we need to do it now, but it seems like it would be better to use `#ifdef` to control when this is included in the compile. Then the client doesn't even see it. Also, I'm not sure that I agree with skipping the checking of return codes. It's true that a failure in the telemetry library shouldn't necessarily be fatal, but I'm concerned that we might be missing some important clues if/when things go wrong. I'll defer to @johannlombardi and other core DAOS developers on this point though.

Comment (jbrosenz) [reply src/cart/crt_context.c:303]: If the DAOS core devs agree to using #ifdef to exclude those calls from the client, that would be a great solution.  If we do get to use #ifdef to exclude them, then it makes it easy to add the checks for errors without generating noise elsewhere.

File: src/iosrv/init.c
Comment (kjacque) [inline src/iosrv/init.c:45]: Just noticed, telemetry_producer.h and _consumer.h don't include telemetry_common.h. If they did, we could include just telemetry_producer.h here.

File: src/include/gurt/telemetry_consumer.h
Comment (kjacque) [inline src/include/gurt/telemetry_consumer.h:2]: Just noticed the contract number in the copyright block isn't current. Should be 8F-30005. Same applies to telemetry_producer.h

File: src/gurt/telemetry.c
Comment (kjacque) [inline src/gurt/telemetry.c:101]: Not a new change, but this line isn't necessary, is it?

File: src/gurt/telemetry.c
Comment (kjacque) [inline src/gurt/telemetry.c:244]: Is the terminology change correct? The variable is called rank, so we should change the variable name too, if the original understanding was wrong.

File: src/gurt/telemetry.c
Comment (kjacque) [inline src/gurt/telemetry.c:2018]: Would it be interesting to log these failure cases? Or can this stuff happen under normal circumstances, such that it would be noisy?

**Review by kjacque — APPROVED** [review]
My comments are minor, and could be addressed in a follow-on.

File: src/gurt/telemetry.c
Comment (jbrosenz) [reply src/gurt/telemetry.c:101]: The file failed a coverity scan and wanted that change.

File: src/gurt/telemetry.c
Comment (jbrosenz) [reply src/gurt/telemetry.c:244]: Server index is correct.  My use of the name 'rank' is not.  I left the variable name as 'rank' for now to minimize changes for this PR.  That said, it's not a big problem to change rank -> srv_idx everywhere, so I'll do that on the next update.

File: src/gurt/telemetry.c
Comment (jbrosenz) [reply src/gurt/telemetry.c:2018]: The failure should not occur often.  There are two reasons why it might fail, 1) system ran out of memory, 2) developer passed in bad arguments.  I don't think that logging it would be noisy, but I wrote it so that it returns the error code so that a developer could read that while developing the code and take necessary action.  If there is consensus on a better approach, I'll retool that as necessary.

**Review by frostedcmos — CHANGES_REQUESTED** [review]
Overall I feel like today tracking number of requests in flight, queue depth, number of aborts on server side is not extremely useful, as those things get tracked on the 'sender' side.

For most cases today 'sender' ends up being client, so information that the server would end up logging/tracking ends up being sporadic server to server communications such as swim pings.

I would like to propose an alternative instead and maybe just start with tracking simple thing on receiver side instead. One such useful metric would be:

Add tracking to: 
 crt_hg.c :: crt_rpc_handler_common(hg_handle_t hg_hdl)
in there tracking can be done on per "crt_ctx" basis.

The information that could be tracked there is number of incoming rpcs destined for that given context, and maybe an aggregated totals of all rpcs received.

We can then later split incoming rpcs based on types (e.g. was it collective rpc vs regular, was it internal cart rpc vs daos vs user). Those types of metrics would help with identifying issues we've already seen in the past

**Comment by mjmac** [comment]
@frostedcmos: Would it be reasonable to just remove the cart RPC metrics for now and add them properly in a subsequent PR? The main goal of this PR is to just get something landed so that we can start exercising the new telemetry library. We picked those RPC metrics from some demo code, so they weren't necessarily intended to be 100% correct. If we remove them, we'll still have some other simple metrics to play with, and we can continue to iterate on adding new metrics with feedback from other developers.

**Comment by frostedcmos** [comment]
> @frostedcmos: Would it be reasonable to just remove the cart RPC metrics for now and add them properly in a subsequent PR? The main goal of this PR is to just get something landed so that we can start exercising the new telemetry library. We picked those RPC metrics from some demo code, so they weren't necessarily intended to be 100% correct. If we remove them, we'll still have some other simple metrics to play with, and we can continue to iterate on adding new metrics with feedback from other developers.

I don't really have any preference on this or objections to this. My main point is that I would rather not land this as is, as the stats wont make much sense being combined as they are. If there is a desire/need to land this asap then perhaps removing rpc metrics for now is the right approach. I'll let @jbrosenz and others decide on this though.

**Comment by jbrosenz** [comment]
I had some work in progress to rework the changes that @frostedcmos suggested.  I see the direction is now to make another PR to add some or most of those later.  I'll put those changes aside and will work on addressing the other feedback and will push an update.

File: src/utils/daos_metrics/daos_metrics.c
Comment (jbrosenz) [reply src/utils/daos_metrics/daos_metrics.c:106]: fixed

File: src/utils/daos_metrics/daos_metrics.c
Comment (jbrosenz) [reply src/utils/daos_metrics/daos_metrics.c:195]: fixed

File: src/utils/daos_metrics/daos_metrics.c
Comment (jbrosenz) [reply src/utils/daos_metrics/daos_metrics.c:50]: Yes, good catch.  Fixed.

File: src/include/gurt/telemetry_consumer.h
Comment (jbrosenz) [reply src/include/gurt/telemetry_consumer.h:2]: Fixed.

File: src/iosrv/init.c
Comment (jbrosenz) [reply src/iosrv/init.c:45]: I'll make that change in a later PR.

File: src/gurt/telemetry.c
Comment (jbrosenz) [reply src/gurt/telemetry.c:244]: I decided to incorporate those changes now.

**Comment by mjmac** [comment]
One known failure:
  - [DAOS-6590](https://jira.hpdd.intel.com/browse/DAOS-6590)

**Comment by mjmac** [comment]
Going to land this to keep momentum. Had positive reviews before needing to merge master.

**Pull request merged**
